pkg/runner/client.go

```diff
@@ -165,16 +165,18 @@ func CreateClient(r prometheus.Registerer, cfg *Config) (*pgclient.Client, error
 		cfg.APICfg.MultiTenancy = multiTenancy
 	}
 
-	if (cfg.DatasetCfg != dataset.Config{}) {
-		if cfg.DatasetConfig != "" {
-			log.Warn("msg", "Ignoring `startup.dataset.config` in favor of the newer `startup.dataset` config option since both were set.")
+	if !cfg.APICfg.ReadOnly {
+		if (cfg.DatasetCfg != dataset.Config{}) {
+			if cfg.DatasetConfig != "" {
+				log.Warn("msg", "Ignoring `startup.dataset.config` in favor of the newer `startup.dataset` config option since both were set.")
+			}
+			err = cfg.DatasetCfg.Apply(conn)
+		} else if cfg.DatasetConfig != "" {
+			err = applyDatasetConfig(conn, cfg.DatasetConfig)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error applying dataset configuration: %w", err)
```

Comment:
I think we can move this logic to a different function, something like:

```
func foo(cfg ) error {
  if cfg.APICfg.ReadOnly {
     return nil     
  }
  if (cfg.DatasetCfg != dataset.Config{}) {
    if cfg.DatasetConfig != "" {
	  log.Warn("msg", "Ignoring `startup.dataset.config` in favor of the newer `startup.dataset` config option since both were set.")
    }
    return cfg.DatasetCfg.Apply(conn)
  }
  if cfg.DatasetConfig != "" {
    return applyDatasetConfig(conn, cfg.DatasetConfig)
  }
  return nil
}

err := foo(cfg)
if err != nil {
  return nil, fmt.Errorf("error applying dataset configuration: %w", err)
}
```

I don't have the brain cells right now to come up with a name for the function. But, WDYT?

Reply:
how about `applyDatasetConfigIfNeeded`

Reply:
How about we take the ReadOnly check and call the function `applyDatasetConfigIfDefined`

```
if !cfg.APICfg.ReadOnly {
  // Applying the dataset config needs a connection with write permissions.
  err := applyDatasetConfigIfDefined(cfg)
  if err != nil {
    return nil, fmt.Errorf("error applying dataset configuration: %w", err)
  }
}

...

func applyDatasetConfigIfDefined(cfg ) error {
  if (cfg.DatasetCfg != dataset.Config{}) {
    if cfg.DatasetConfig != "" {
	  log.Warn("msg", "Ignoring `startup.dataset.config` in favor of the newer `startup.dataset` config option since both were set.")
    }
    return cfg.DatasetCfg.Apply(conn)
  }
  if cfg.DatasetConfig != "" {
    return applyDatasetConfig(conn, cfg.DatasetConfig)
  }
  return nil
}
```

Reply:
sounds good! adjusted accord to the suggestion